### PR TITLE
RaycastPerceptionComponent custom editor

### PR DESCRIPTION
**Pull request by chriselion**
### Proposed change(s)

* Make RayPerceptionSensorComponent fields private. *Most* of them now have public getters and setters. Ones that affect the observation shape of the sensor (num rays, tags, and stacks) properties are not public. This allows changing of some sensor values at runtime (e.g. someone asked for changing the ray length as part of generalization).
* Add a custom editor for RayPerceptionSensorComponent. Parameters that aren't allowed to change during training/inference get disabled in play mode.
* Make debug drawing for RayPerceptionSensorComponent work outside of play mode. Slight behavior change in that only selected Agent(s) are drawn.

### Useful links (Github issues, JIRA tickets, ML-Agents forum threads etc.)
https://jira.unity3d.com/browse/MLA-688 
https://jira.unity3d.com/browse/MLA-561

https://github.com/Unity-Technologies/ml-agents/issues/3412
https://forum.unity.com/threads/no-visible-gizmos-for-ray-perception-sensor-component-3d.824529/
https://forum.unity.com/threads/access-rayperceptionsensor-components.823107/

### Types of change(s)

- [ ] Bug fix
- [x] New feature
- [x] Code refactor
- [x] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added updated the changelog (if applicable)
- [ ] I have added necessary documentation (if applicable)
- [ ] I have updated the migration guide (if applicable)

### CODE REVIEW

**Comment by chriselion** [comment]
out

File: com.unity.ml-agents/Runtime/Sensor/RayPerceptionSensorComponentBase.cs [hunk 1]
```diff
@@ -1,42 +1,112 @@
...
 namespace MLAgents
 {
     public abstract class RayPerceptionSensorComponentBase : SensorComponent
     {
-        public string sensorName = "RayPerceptionSensor";
+        [HideInInspector]
+        [SerializeField]
+        [FormerlySerializedAs("sensorName")]
+        string m_SensorName = "RayPerceptionSensor";
+        internal string sensorName
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Sensor/RayPerceptionSensorComponentBase.cs]: TODO change the `internal` ones to have public getters and internal setters.

File: com.unity.ml-agents/Runtime/Sensor/RayPerceptionSensorComponent3D.cs [hunk 2]
```diff
@@ -1,19 +1,35 @@
...
 
 namespace MLAgents
 {
     [AddComponentMenu("ML Agents/Ray Perception Sensor 3D", (int)MenuGroup.Sensors)]
     public class RayPerceptionSensorComponent3D : RayPerceptionSensorComponentBase
     {
-        [Header("3D Properties", order = 100)]
+        [HideInInspector]
+        [SerializeField]
+        [FormerlySerializedAs("startVerticalOffset")]
```
Comment (surfnerd) [inline com.unity.ml-agents/Runtime/Sensor/RayPerceptionSensorComponent3D.cs]: nit: attributes should go on the same line if they fit.

File: com.unity.ml-agents/Runtime/Sensor/RayPerceptionSensorComponentBase.cs [hunk 3]
```diff
@@ -1,42 +1,112 @@
...
+        {
+            get => m_SensorName;
+            set => m_SensorName = value;
+        }
 
+        //[HideInInspector]
```
Comment (surfnerd) [inline com.unity.ml-agents/Runtime/Sensor/RayPerceptionSensorComponentBase.cs]: nit: remove

Comment (chriselion) [reply com.unity.ml-agents/Runtime/Sensor/RayPerceptionSensorComponentBase.cs]: Will do. I'm not quite clear why we have this in the first place for things like this:
https://github.com/Unity-Technologies/ml-agents/blob/9450d3fc0dda4547a14c5ed1b7e13fc6e3a15413/com.unity.ml-agents/Runtime/Policy/BehaviorParameters.cs#L35-L37
but then expose it in the editor: https://github.com/Unity-Technologies/ml-agents/blob/9450d3fc0dda4547a14c5ed1b7e13fc6e3a15413/com.unity.ml-agents/Editor/BehaviorParametersEditor.cs#L28

I had to remove it here to get the list to expand properly.

**Comment by chriselion** [comment]
>  I wonder if these would be helpful in a player build as well.

Do you mean the gizmo drawing? What's the typical setup?

**Pull request merged**